### PR TITLE
Executing JavaScript callback on every scenario

### PR DESCRIPTION
**Pull request by sagrawal31**
## Problem

As described in the issue https://github.com/chemerisuk/cordova-plugin-firebase-dynamiclinks/issues/54#issuecomment-562826134

Hi @chemerisuk 

This is an improvement to the plugin to execute the callback on every scenario. This is related to #54

```javascript
cordova.plugins.firebase.dynamiclinks.onDynamicLink(function(data) {
    console.log("App launched:", data);
});
```

## Current Behaviour

Currently, the following thing happens:

Scenario | Above callback executes
----|-----
The app is launched directly from the app icon | No
The app is launched from `APP_DOMAIN` URL | No
The app is launched from deep link URL `PAGE_LINK_DOMAIN` | Yes

## New Behaviour

Now, after we do this change, the following will happen:

Scenario | Above callback executes | Callback example
----|---|---
The app is launched directly from the app icon | Yes | image
The app is launched from `APP_DOMAIN` URL | Yes | image
The app is launched from deep link URL `PAGE_LINK_DOMAIN` | Yes | image

## Breaking Change

For existing developers of this plugin, if an app is not launched from the deep link URL, their callback will still be invoked.

## Workaround/Solution

Existing developers can add a simple check in their code like:

```javascript
cordova.plugins.firebase.dynamiclinks.onDynamicLink(function(data) {
    if (data.hasDeepLink) {
        // there existing code
    }
});
```

## Pending

Similar changes for iOS.

## Benefits of this change

1. Cordova developers don't have to be dependent on other plugins to figure out a way that how an app is launched.
2. Since the callback will be executed on every app launch scenario, the developer doesn't have to put any kind of wait/sleep timings just to see how the app was launched.
3. Existing developers using this plugin even don't have to change anything major in their code except one simple `if` condition.

### CODE REVIEW

**Comment by sagrawal31** [comment]
@chemerisuk awaiting your positive reply :smile:

**Comment by chemerisuk** [comment]
@sagrawal31 I have verified your use cases. If dynamic link is correct, `APP_DOMAIN` links work as well. For the 1st scenario - why the callback should be triggered on app start? The goal for `onDynamicLink` is to return dynamic link string so on client side we can do appropriate actions. When we can't get the dynamic link (or url is broken etc.) - no reason to trigger callback.

In your case I guess you didn't setup/register dynamic link domain. Try to call `createShortDynamicLink` to get details if any.

**Comment by sagrawal31** [comment]
> For the 1st scenario - why the callback should be triggered on app start?

@chemerisuk I have explained it in my issue comment. Let me rephrase it. Basically, when an app is launched, we should be able to know that whether the app is launched through a deeplink/dynamic link and whether the code should wait for the `onDynamicLink` callback to get executed in order to perform some activity. For example, consider this dynamic link URL: https://example.page.link/?apn=com.example&isi=1450713709&ibi=com.example&link=https%3A%2F%2Fexample.com%3Fxyz%3D1%2526foo%3Dbar

If the app wants to log an event (app launch event) in the server where it wants to pass-on some parameters (`xyz` & `foo` in this case) to the server if the app is launched from dynamic link otherwise do not hit the server.

So this MR will tell the mobile app code by checking `hasDeepLink` (which will be `true`/`false`) that if the app was launched from directly pressing the icon (the 1st scenario or where there is no pending dynamic link data) or from the dynamic link. If `hasDeepLink === true` then I can parse the `deepLink` value from the callback (which will be https://example.com?xyz=1%26foo=bar in this case) and can invoke my backend server API by passing-on those values.

Otherwise, when `hasDeepLink === false`, then the mobile app knows that there is no pending dynamic link data and can move forward.

> When we can't get the dynamic link (or url is broken etc.) - no reason to trigger callback.

Totally agree with you, but when it comes to analytics and advertisment, developers/business wants to know how & from where their apps are being launched and if they are using Firebase dynamiclink, the code should be able to know if their is any pending dynamic link data or not.

This MR modified the current callback behaviour. May be we can change this to something else which only gives us the `true`/`false`. For example:

```javascript
cordova.plugins.firebase.dynamiclinks.hasDynamicLink(function(doWeHavePendingDL) {
    console.log("Is pending dynamiclink: ", doWeHavePendingDL === true);
});
```
But this might complicate things like:

```javascript
cordova.plugins.firebase.dynamiclinks.hasDynamicLink(function(doWeHavePendingDL) {

    if (doWeHavePendingDL) {
        cordova.plugins.firebase.dynamiclinks.onDynamicLink(function(data) {
            console.log("App launched:", data);
        });
    }
});
```

> In your case I guess you didn't setup/register dynamic link domain. Try to call createShortDynamicLink to get details if any.

Dynamic link is properly setup @chemerisuk I have verified all the ways.

**Comment by sagrawal31** [comment]
On second thought after my today's 3-4 hours of research and testing on a live Google Play app, actually, those 3 scenarios can be merged into one that it doesn't matter how the app was launched, we get a callback which gives us a value (in `hasDeepLink` to `true`/`false`)  which says that if there is any pending dynamic link data or not.

I tested this scenario in the following way:

1. My phone doesn't have an app `Example App`
2. The Firebase dynamic link URL was invoked in the URL which took me to the Google Play app.
3. From there, I downloaded & installed the app.
4. Now, instead of hitting the `Continue` button in Google Play, I launched the newly installed app directly from the app icon.
5. And I received the following data:

Screenshot 2019-12-12 at 5 37 28 PM

Basically, even I launched the app from the icon (where it was installed through a dynamic link), I got to know that there is a pending dynamic link data.

**Comment by chemerisuk** [comment]
@sagrawal31 I see, so what you are suggesting is to modify plugin behavior for use cases you've described... Frankly it doesn't sound right for me. The purpose of the plugin is to handle **Firebase Dynamic Links**. For anything else it's better to write an appropriate plugin.

I do see the source of confusion in variable `APP_DOMAIN`.  It's here for historical reasons (in the past there were no *.page.link domains), but now it should probably be removed

**Comment by chemerisuk** [comment]
@sagrawal31 current implementation handles the use case you've described in the last comment.

**Comment by sagrawal31** [comment]
> so what you are suggesting is to modify plugin behavior for use cases you've described

😄 Truly speaking, not really. I thought of this because, as of now, we don't have any way in JavaScript code to know that if there is a pending dynamic link data if we are using this plugin.

So probably, I can modify my MR to make it simple that really doesn't make you feel that something is being changed majorly. 😛 I agree, that this might be a breaking change (as I explained above) for existing developers who are using this plugin but the plugin should give a way to tell that if there is any pending dynamic link data.

> I do see the source of confusion in variable `APP_DOMAIN`

True. At first, I was confused too because of this but later on, understood the different points.

> current implementation handles the use case you've described in the last comment.

Yes yes 🙃🙂 I'm aware of that. I just wanted to make sure that my changes are not breaking any sort of existing functionality so I went the extra mile.

**Comment by sagrawal31** [comment]
Further thoughts please @chemerisuk?

**Comment by chemerisuk** [comment]
@sagrawal31 for your use cases it's better to create a separate plugin. I want to keep simple this plugin, only focusing on FDL support.

**Comment by sagrawal31** [comment]
Thank you for your response @chemerisuk 

> I want to keep simple this plugin, only focusing on FDL support.

I do share the same thoughts and I'm just trying to add a small way to know in advance that do we have any pending dynamic link data.

To discuss more with you and align with your intention, I have modified the code a little bit (not tested yet). Now, the **original functionality is maintained** i.e. the following callback will only execute if there is any pending data.

```javascript
cordova.plugins.firebase.dynamiclinks.onDynamicLink(function(data) {
    console.log("App launched:", data);
});
```
If someone (like us 😄) needs callback even there is no deep link data, they can pass `true` as the last parameter.

```javascript
cordova.plugins.firebase.dynamiclinks.onDynamicLink(function(data) {
    console.log("App launched:", data);
}, function(error) {
    console.log("FDL failed:", error);
}, true);
```

This way the original functionality will be maintained and there will no other logic then the Firebase DynamicLink handling.

Your further thoughts, please!

(By the way, I have also removed one extra code of passing `deepLinkURL`)

**Comment by chemerisuk** [comment]
`APP_DOMAIN ` was removed in the latest version.
